source/Timer.cpp

```diff
@@ -0,0 +1,307 @@
+/* Timer.cpp
+Copyright (c) 2023 by Timothy Collett
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Timer.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "GameData.h"
+#include "Logger.h"
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+#include "Ship.h"
+#include "StellarObject.h"
+#include "UI.h"
+
+using namespace std;
+
+Timer::Timer(const DataNode &node, const Mission *mission)
+{
+	Load(node, mission);
+}
+
+
+
+void Timer::Load(const DataNode &node, const Mission *mission)
+{
+	if(node.Size() > 1)
+		name = node.Token(1);
+
+	this->mission = mission;
+
+	timeToWait = base + Random::Int(rand);
+
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "time" && child.Size() > 1)
+		{
+			base = static_cast<int64_t>(child.Value(1));
+			if(child.Size() > 2)
+				rand = static_cast<uint32_t>(child.Value(2));
+		}
+		else if(child.Token(0) == "idle")
+			requireIdle = true;
+		else if(child.Token(0) == "optional")
+			optional = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity")
```

Comment:
Hm. Are there other places that take a LocationFilter that _must_ be a Planet, not a System? I suppose either way, this could do that—it would just have to fail to load the Timer altogether if the LocationFilter doesn't resolve to one or more Planets.

Reply:
> it would just have to fail to load the Timer altogether

Not necessarily, it can just do nothing, like in other places that allow a LocationFilter.